openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,551 @@
+			if r == coreapi.ResourceCPU && request.IsZero() {
+				continue
+			}
+
+			if !limit.Equal(request) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func isBestEffort(containers []coreapi.Container) bool {
+	for _, c := range containers {
+		// only memory and CPU resources are relevant to decide pod QoS class
+		for _, r := range []coreapi.ResourceName{coreapi.ResourceMemory, coreapi.ResourceCPU} {
+			limit := c.Resources.Limits[r]
+			request := c.Resources.Requests[r]
+
+			if !limit.IsZero() || !request.IsZero() {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func getPodQoSClass(containers []coreapi.Container) coreapi.PodQOSClass {
+	if isGuaranteed(containers) {
+		return coreapi.PodQOSGuaranteed
+	}
+
+	if isBestEffort(containers) {
+		return coreapi.PodQOSBestEffort
+	}
+
+	return coreapi.PodQOSBurstable
+}
+
+func isManagementResourceAvailableForAllNodes(nodes []*corev1.Node, workloadType string) error {
+	managedResource := fmt.Sprintf("%s.%scores", workloadType, podWorkloadAnnotationPrefix)
+	for _, node := range nodes {
+		if v, ok := node.Status.Allocatable[corev1.ResourceName(managedResource)]; !ok || v.IsZero() {
+			return fmt.Errorf("the node %q does not have resource %q", node.Name, managedResource)
+		}
+	}
+
+	return nil
+}
+
+func doesNamespaceAllowWorkloadType(annotations map[string]string, workloadType string) bool {
+	v, found := annotations[namespaceAllowedAnnotation]
+	if !found {
+		return false
+	}
+
+	for _, t := range strings.Split(v, ",") {
+		if workloadType == t {
+			return true
+		}
+	}
+
+	return false
+}
+
+func getWorkloadType(annotations map[string]string) (string, error) {
+	var workloadAnnotationsKeys []string
+	for k := range annotations {
+		if strings.HasPrefix(k, podWorkloadAnnotationPrefix) {
+			workloadAnnotationsKeys = append(workloadAnnotationsKeys, k)
+		}
+	}
+
+	// no workload annotation is specified under the pod
+	if len(workloadAnnotationsKeys) == 0 {
+		return "", nil
+	}
+
+	// more than one workload annotation exists under the pod and we do not support different workload types
+	// under the same pod
+	if len(workloadAnnotationsKeys) > 1 {
+		return "", fmt.Errorf("the pod can not have more than one workload annotations")
+	}
+
+	workloadType := strings.TrimPrefix(workloadAnnotationsKeys[0], podWorkloadAnnotationPrefix)
+	if len(workloadType) == 0 {
+		return "", fmt.Errorf("the workload annotation key should have format %s<workload_type>, when <workload_type> is non empty string", podWorkloadAnnotationPrefix)
+	}
+
+	return workloadType, nil
+}
+
+func getWorkloadAnnotationEffect(workloadAnnotationKey string) (string, error) {
+	managementAnnotationValue := map[string]string{}
+	if err := json.Unmarshal([]byte(workloadAnnotationKey), &managementAnnotationValue); err != nil {
+		return "", fmt.Errorf("failed to parse %q annotation value: %v", workloadAnnotationKey, err)
+	}
+
+	effect, ok := managementAnnotationValue[podWorkloadAnnotationEffect]
```

Comment:
you need to enforce all the possible keys as well.

Reply:
I added a check to guarantee that the annotation value map has only one key.